File: crates/gateway-http-anthropic/src/tool_arg_policy.rs
#![forbid(unsafe_code)]

use serde_json::Map;
use serde_json::Value;

use gateway_backend_codex::tool_calls::normalize_json_object_string;
use gateway_backend_codex::types::CodexToolCallKind;

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct PolicyEdit {
    pub(crate) field: &'static str,
    pub(crate) action: &'static str,
    pub(crate) reason: &'static str,
}

pub(crate) struct ToolArgContext<'a> {
    pub(crate) tool_name: &'a str,
}

pub(crate) fn apply_policies(
    ctx: &ToolArgContext<'_>,
    args: &mut Map<String, Value>,
) -> Vec<PolicyEdit> {
    let mut edits = Vec::new();
    edits.extend(agent_policy(ctx, args));
    edits.extend(read_policy(ctx, args));
    edits
}

pub(crate) fn sanitized_tool_args_for_kind(
    tool_name: &str,
    kind: CodexToolCallKind,
    buf: &str,
) -> Result<(Map<String, Value>, Vec<PolicyEdit>), String> {
    let mut args = parse_tool_args_object_for_kind(kind, buf)?;
    let ctx = ToolArgContext { tool_name };
    let edits = apply_policies(&ctx, &mut args);
    Ok((args, edits))
}

fn validate_tool_args_json_object(buf: &str) -> Result<(), String> {
    let trimmed = buf.trim();
    if trimmed.is_empty() {
        return Ok(());
    }
    let value: Value = serde_json::from_str(trimmed)
        .map_err(|error| format!("tool_use.input is not valid JSON: {error}"))?;
    if !value.is_object() {
        return Err("tool_use.input must be a JSON object".to_string());
    }
    Ok(())
}

fn parse_tool_args_object(buf: &str) -> Result<Map<String, Value>, String> {
    let trimmed = buf.trim();
    validate_tool_args_json_object(trimmed)?;
    if trimmed.is_empty() {
        return Ok(Map::new());
    }
    let value: Value = serde_json::from_str(trimmed)
        .map_err(|error| format!("tool_use.input is not valid JSON: {error}"))?;
    let obj = value
        .as_object()
        .cloned()
        .ok_or_else(|| "tool_use.input must be a JSON object".to_string())?;
    Ok(obj)
}

fn parse_tool_args_object_for_kind(
    kind: CodexToolCallKind,
    buf: &str,
) -> Result<Map<String, Value>, String> {
    if kind != CodexToolCallKind::Custom {
        return parse_tool_args_object(buf);
    }

    let trimmed = buf.trim();
    if trimmed.is_empty() {
        return Ok(Map::new());
    }
    if let Ok(obj) = parse_tool_args_object(trimmed) {
        return Ok(obj);
    }

    let normalized = normalize_json_object_string(trimmed, "input");
    parse_tool_args_object(&normalized)
}

fn agent_policy(ctx: &ToolArgContext<'_>, args: &mut Map<String, Value>) -> Vec<PolicyEdit> {
    if ctx.tool_name != "Agent" || !args.contains_key("isolation") {
        return Vec::new();
    }

    args.remove("isolation");
    vec![PolicyEdit {
        field: "isolation",
        action: "remove",
        reason: "gateway should not force worktree isolation for Agent calls",
    }]
}

fn read_policy(ctx: &ToolArgContext<'_>, args: &mut Map<String, Value>) -> Vec<PolicyEdit> {
    if ctx.tool_name != "Read" {
        return Vec::new();
    }

    let mut edits = Vec::new();
    let pages = args.get("pages");
    if pages.is_none() {
        return edits;
    }

    let file_path = args.get("file_path").and_then(Value::as_str);
    let is_pdf = file_path.is_some_and(|p| p.to_ascii_lowercase().ends_with(".pdf"));

    let pages_empty = pages.and_then(Value::as_str).is_some_and(str::is_empty);

    if pages_empty {
        args.remove("pages");
        edits.push(PolicyEdit {
            field: "pages",
            action: "remove",
            reason: "empty string is invalid; omit pages unless reading a PDF",
        });
        return edits;
    }

    if !is_pdf {
        args.remove("pages");
        edits.push(PolicyEdit {
            field: "pages",
            action: "remove",
            reason: "pages only applies to PDF reads",
        });
    }

    edits
}

#[cfg(test)]
mod tests {
    use super::*;

    fn ctx(tool_name: &str) -> ToolArgContext<'_> {
        ToolArgContext { tool_name }
    }

    #[test]
    fn read_drops_empty_pages() {
        let mut args = Map::from_iter([
            (
                "file_path".to_string(),
                Value::String("/tmp/a.txt".to_string()),
            ),
            ("pages".to_string(), Value::String(String::new())),
        ]);
        let edits = apply_policies(&ctx("Read"), &mut args);
        assert!(args.get("pages").is_none());
        assert_eq!(edits.len(), 1);
    }

    #[test]
    fn read_drops_pages_for_non_pdf() {
        let mut args = Map::from_iter([
            (
                "file_path".to_string(),
                Value::String("/tmp/a.txt".to_string()),
            ),
            ("pages".to_string(), Value::String("1-2".to_string())),
        ]);
        let edits = apply_policies(&ctx("Read"), &mut args);
        assert!(args.get("pages").is_none());
        assert_eq!(edits.len(), 1);
    }

    #[test]
    fn read_keeps_pages_for_pdf() {
        let mut args = Map::from_iter([
            (
                "file_path".to_string(),
                Value::String("/tmp/a.PDF".to_string()),
            ),
            ("pages".to_string(), Value::String("1-2".to_string())),
        ]);
        let edits = apply_policies(&ctx("Read"), &mut args);
        assert!(args.get("pages").is_some());
        assert!(edits.is_empty());
    }

    #[test]
    fn agent_drops_isolation() {
        let mut args = Map::from_iter([
            (
                "description".to_string(),
                Value::String("Research files".to_string()),
            ),
            (
                "prompt".to_string(),
                Value::String("Inspect relevant files".to_string()),
            ),
            (
                "isolation".to_string(),
                Value::String("worktree".to_string()),
            ),
        ]);

        let edits = apply_policies(&ctx("Agent"), &mut args);

        assert!(args.get("isolation").is_none());
        assert_eq!(edits.len(), 1);
        assert_eq!(edits[0].field, "isolation");
    }

    #[test]
    fn tool_args_validation_requires_object() {
        let err = validate_tool_args_json_object("[]").expect_err("should reject non-object");
        assert_eq!(err, "tool_use.input must be a JSON object");
    }

    #[test]
    fn parse_tool_args_object_accepts_empty_as_object() {
        let obj = parse_tool_args_object("").expect("empty ok");
        assert!(obj.is_empty());
    }

    #[test]
    fn custom_tool_args_wrap_raw_input() {
        let (args, edits) =
            sanitized_tool_args_for_kind("apply_patch", CodexToolCallKind::Custom, "raw patch")
                .expect("sanitize");
        assert_eq!(args.get("input").and_then(Value::as_str), Some("raw patch"));
        assert!(edits.is_empty());
    }
}

File: golang_port/core/domain/translator/tool_arg_policy.go
package translator

import (
	"encoding/json"
	"fmt"
	"strings"
)

// ToolCallKind mirrors the backend tool-call kinds a translator needs to
// know about to parse a raw tool-call argument buffer (port of
// gateway-backend-codex::types::CodexToolCallKind). Kept as an independent
// domain-level type rather than importing the codex impl package, so this
// policy stays backend-agnostic per the translator port's extend-via-
// composition design (ARCHITECTURE_v2.md).
// Values match the canonical wire-format strings from
// crates/gateway-backend-codex/src/types.rs::CodexToolCallKind::as_str().
type ToolCallKind string

const (
	ToolCallKindFunction   ToolCallKind = "function_call"
	ToolCallKindCustom     ToolCallKind = "custom_tool_call"
	ToolCallKindToolSearch ToolCallKind = "tool_search_call"
	ToolCallKindLocalShell ToolCallKind = "local_shell_call"
)

// PolicyEdit records one tool-arg mutation applied by ApplyPolicies, for
// observability/audit. Port of tool_arg_policy.rs PolicyEdit.
type PolicyEdit struct {
	Field  string
	Action string
	Reason string
}

// ToolArgContext carries the inputs a policy needs to decide whether to
// act. Port of tool_arg_policy.rs ToolArgContext.
type ToolArgContext struct {
	ToolName string
}

// ApplyPolicies runs all tool-arg policies against args in place, returning
// the edits applied. Port of tool_arg_policy.rs apply_policies.
func ApplyPolicies(ctx ToolArgContext, args map[string]any) []PolicyEdit {
	edits := make([]PolicyEdit, 0)
	edits = append(edits, agentPolicy(ctx, args)...)
	edits = append(edits, readPolicy(ctx, args)...)
	return edits
}

// SanitizedToolArgsForKind parses buf into a JSON object appropriate for
// kind, then applies ApplyPolicies. Port of tool_arg_policy.rs
// sanitized_tool_args_for_kind.
func SanitizedToolArgsForKind(toolName string, kind ToolCallKind, buf string) (map[string]any, []PolicyEdit, error) {
	args, err := parseToolArgsObjectForKind(kind, buf)
	if err != nil {
		return nil, nil, err
	}
	ctx := ToolArgContext{ToolName: toolName}
	edits := ApplyPolicies(ctx, args)
	return args, edits, nil
}

func validateToolArgsJSONObject(buf string) error {
	trimmed := strings.TrimSpace(buf)
	if trimmed == "" {
		return nil
	}
	var value any
	if err := json.Unmarshal([]byte(trimmed), &value); err != nil {
		return fmt.Errorf("tool_use.input is not valid JSON: %w", err)
	}
	if _, ok := value.(map[string]any); !ok {
		return fmt.Errorf("tool_use.input must be a JSON object")
	}
	return nil
}

func parseToolArgsObject(buf string) (map[string]any, error) {
	trimmed := strings.TrimSpace(buf)
	if err := validateToolArgsJSONObject(trimmed); err != nil {
		return nil, err
	}
	if trimmed == "" {
		return map[string]any{}, nil
	}
	var value any
	if err := json.Unmarshal([]byte(trimmed), &value); err != nil {
		return nil, fmt.Errorf("tool_use.input is not valid JSON: %w", err)
	}
	obj, ok := value.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("tool_use.input must be a JSON object")
	}
	return obj, nil
}

func parseToolArgsObjectForKind(kind ToolCallKind, buf string) (map[string]any, error) {
	if kind != ToolCallKindCustom {
		return parseToolArgsObject(buf)
	}

	trimmed := strings.TrimSpace(buf)
	if trimmed == "" {
		return map[string]any{}, nil
	}
	if obj, err := parseToolArgsObject(trimmed); err == nil {
		return obj, nil
	}

	normalized := normalizeJSONObjectString(trimmed, "input")
	return parseToolArgsObject(normalized)
}

// normalizeJSONObjectString parses raw as JSON; if it decodes to an object,
// that object (re-encoded) is returned. Otherwise raw is wrapped as
// {fallbackField: raw}. Empty/whitespace-only raw becomes "{}". Port of
// gateway-backend-codex::tool_calls::normalize_json_object_string, kept
// local to this package for the same backend-agnostic reason as
// ToolCallKind above.
func normalizeJSONObjectString(raw, fallbackField string) string {
	trimmed := strings.TrimSpace(raw)
	if trimmed == "" {
		return "{}"
	}
	var value any
	if err := json.Unmarshal([]byte(trimmed), &value); err != nil {
		encoded, marshalErr := json.Marshal(map[string]any{fallbackField: raw})
		if marshalErr != nil {
			return "{}"
		}
		return string(encoded)
	}
	objectValue := value
	if _, ok := value.(map[string]any); !ok {
		objectValue = map[string]any{fallbackField: value}
	}
	encoded, err := json.Marshal(objectValue)
	if err != nil {
		return "{}"
	}
	return string(encoded)
}

func agentPolicy(ctx ToolArgContext, args map[string]any) []PolicyEdit {
	if ctx.ToolName != "Agent" {
		return nil
	}
	if _, ok := args["isolation"]; !ok {
		return nil
	}
	delete(args, "isolation")
	return []PolicyEdit{{
		Field:  "isolation",
		Action: "remove",
		Reason: "gateway should not force worktree isolation for Agent calls",
	}}
}

func readPolicy(ctx ToolArgContext, args map[string]any) []PolicyEdit {
	if ctx.ToolName != "Read" {
		return nil
	}

	edits := make([]PolicyEdit, 0)
	pages, hasPages := args["pages"]
	if !hasPages {
		return edits
	}

	filePath, _ := args["file_path"].(string)
	isPDF := strings.HasSuffix(strings.ToLower(filePath), ".pdf")

	pagesStr, pagesIsString := pages.(string)
	pagesEmpty := pagesIsString && pagesStr == ""

	if pagesEmpty {
		delete(args, "pages")
		edits = append(edits, PolicyEdit{
			Field:  "pages",
			Action: "remove",
			Reason: "empty string is invalid; omit pages unless reading a PDF",
		})
		return edits
	}

	if !isPDF {
		delete(args, "pages")
		edits = append(edits, PolicyEdit{
			Field:  "pages",
			Action: "remove",
			Reason: "pages only applies to PDF reads",
		})
	}

	return edits
}
